typing_extensions/src_py3/typing_extensions.py

```diff
@@ -620,3 +664,300 @@ def new_type(x):
+                raise TypeError(
+                    "Parameters to %r[...] must all be type variables" % self)
+            if len(set(params)) != len(params):
+                raise TypeError(
+                    "Parameters to %r[...] must all be unique" % self)
+            tvars = params
+            args = params
+        elif self in (Tuple, Callable):
+            tvars = _type_vars(params)
+            args = params
+        elif self.__origin__ in (Generic, Protocol):
+            raise TypeError("Cannot subscript already-subscripted %s" %
+                            repr(self))
+        else:
+            _check_generic(self, params)
+            tvars = _type_vars(params)
+            args = params
+
+        prepend = (self,) if self.__origin__ is None else ()
+        return self.__class__(self.__name__,
+                              prepend + self.__bases__,
+                              _no_slots_copy(self.__dict__),
+                              tvars=tvars,
+                              args=args,
+                              origin=self,
+                              extra=self.__extra__,
+                              orig_bases=self.__orig_bases__)
+
+    if OLD_GENERICS:
+        del __getitem__
+
+
+if NO_PROTOCOL:
+    del _ProtocolMeta
+else:
+    class Protocol(metaclass=_ProtocolMeta):
+        """Base class for protocol classes. Protocol classes are defined as::
+
+          class Proto(Protocol):
+              def meth(self) -> int:
+                  ...
+
+        Such classes are primarily used with static type checkers that recognize
+        structural subtyping (static duck-typing), for example::
+
+          class C:
+              def meth(self) -> int:
+                  return 0
+
+          def func(x: Proto) -> int:
+              return x.meth()
+
+          func(C())  # Passes static type check
+
+        See PEP 544 for details. Protocol classes decorated with @typing_extensions.runtime
+        act as simple-minded runtime protocols that checks only the presence of
+        given attributes, ignoring their type signatures.
+
+        Protocol classes can be generic, they are defined as::
+
+          class GenProto({bases}):
+              def meth(self) -> T:
+                  ...
+        """
+
+        __slots__ = ()
+        _is_protocol = True
+
+        def __new__(cls, *args, **kwds):
+            if _gorg(cls) is Protocol:
+                raise TypeError("Type Protocol cannot be instantiated; "
+                                "it can be used only as a base class")
+            if OLD_GENERICS:
+                return _generic_new(_next_in_mro(cls), cls, *args, **kwds)
+            return _generic_new(cls.__next_in_mro__, cls, *args, **kwds)
+
+    Protocol.__doc__ = Protocol.__doc__.format(bases="Protocol, Generic[T]" if
+                                               OLD_GENERICS else "Protocol[T]")
+
+
+def runtime(cls):
```

Comment:
I still don't like the `@runtime` name. Maybe `@runtime_checkable`? The longer this is the better, to discourage people from using it without understanding the consequences.